### ISSUE TITLE
Nullability inference leads to false positive CS8619

### ISSUE THREAD

**Issue by AroglDarthu** [dotnet/roslyn#51601]
**Version Used**: 
Visual Studio 16.8.6

**Steps to Reproduce**:
- LangVersion latest (C#9)
- Nullable enable

``` CSharp
using System;
using System.Collections.Generic;
using System.Linq;

namespace ConsoleApp3
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
        }

        public IEnumerable<(T?, T)> Convert1<T>(IEnumerable<(T Key, T? Value)> source)
        {
            // No warnings, as expected.
            var result = source.Select(kvp => (kvp.Value, kvp.Key));
            return result;
        }

        public IEnumerable<(T?, T)> Convert2<T>(IEnumerable<Blah<T, T?>> source)
        {
            var result = source.Select(b => (b.Value, b.Key));
            // The following line gives a false positive on CS8619
            return result;
        }

        public IEnumerable<(T?, T)> Convert3<T>(IEnumerable<Blah<T, T?>> source)
        {
            // Without any statements before the return, the warning does not appear...
            return source.Select(b => (b.Value, b.Key));
        }

        public IEnumerable<(T?, T)> Convert4<T>(IEnumerable<Blah<T, T?>> source)
        {
            int x = 3;
            // The following line gives a false positive on CS8619
            return source.Select(b => (b.Value, b.Key));
        }

        public IEnumerable<Blah<T?, T>> Convert5<T>(IEnumerable<Blah<T, T?>> source)
        {
            var result = source.Select(kvp => CreateBlah(kvp.Value, kvp.Key));
            // The following line gives a false positive on CS8619
            return result;
        }

        private static Blah<T1, T2> CreateBlah<T1, T2>(T1 key, T2 value)
        {
            return new(key, value);
        }
    }

    public sealed class Blah<T1, T2>
    {
        public Blah(T1 key, T2 value)
        {
            Key = key;
            Value = value;
        }

        public T1 Key { get; }

        public T2 Value { get; }
    }
}
```

**Expected Behavior**:
No warnings on the return in method `Convert2` and `Convert4` and `Convert5`.

**Actual Behavior**:
The return warns with the following message:
`Warning	CS8619	Nullability of reference types in value of type 'IEnumerable<(T Value, T Key)>' doesn't match target type 'IEnumerable<(T?, T)>'.`

**Comment by RikkiGibson** [comment]
Could you please let us know if this is still the same in 16.9 which released today.

**Comment by AroglDarthu** [comment]
I am both happy and sad now... I updated my VS yesterday to 16.8.6, apparently just before the 16.9 release. The latter indeed does not show these issues!
🤦
